### PR TITLE
Topic/v1.10/mpi win create dynamic

### PR DESCRIPTION
**Pull request by ggouaillardet**
Add missing Fortran bindings for
- MPI_Win_attach
- MPI_Win_detach
- MPI_Win_create_dynamic

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/915/ for details.

**Comment by ggouaillardet** [comment]
:bot:milestone:v1.10.1
:bot:label:enhancement
:bot:label:bug
:bot:label: @jsquyres

File: ompi/mpi/fortran/use-mpi-f08/mpi-f-interfaces-bind.h [hunk 1]
```diff
@@ -2391,6 +2393,34 @@ subroutine ompi_win_create_f(base,size,disp_unit,info,comm,win,ierror) &
...
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   INTEGER(MPI_ADDRESS_KIND), INTENT(IN) :: size
+   INTEGER, INTENT(OUT) :: win
+   INTEGER, INTENT(OUT) :: ierror
+end subroutine ompi_win_attach_f
+
+subroutine ompi_win_detach_f(win,base,ierror) &
+   BIND(C, name="ompi_win_detach_f")
+   use :: mpi_f08_types, only : MPI_ADDRESS_KIND
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f-interfaces-bind.h:2417]: Don't need this MPI_ADDRESS_KIND

File: ompi/mpi/fortran/use-mpi-f08/mpi-f-interfaces-bind.h [hunk 2]
```diff
@@ -2391,6 +2393,34 @@ subroutine ompi_win_create_f(base,size,disp_unit,info,comm,win,ierror) &
...
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   INTEGER, INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f-interfaces-bind.h]: Shouldn't `win` be an IN?

File: ompi/mpi/fortran/use-mpi-f08/mpi-f-interfaces-bind.h [hunk 3]
```diff
@@ -2391,6 +2393,34 @@ subroutine ompi_win_create_f(base,size,disp_unit,info,comm,win,ierror) &
...
+   INTEGER, INTENT(OUT) :: ierror
+end subroutine ompi_win_create_dynamic_f
+
+subroutine ompi_win_attach_f(win,base,size,ierror) &
+   BIND(C, name="ompi_win_attach_f")
+   use :: mpi_f08_types, only : MPI_ADDRESS_KIND
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   INTEGER(MPI_ADDRESS_KIND), INTENT(IN) :: size
+   INTEGER, INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f-interfaces-bind.h]: `win` should be an IN

File: ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces.F90 [hunk 4]
```diff
@@ -3219,6 +3221,48 @@ subroutine MPI_Win_create_f08(base,size,disp_unit,info,comm,win,ierror)
...
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
+   implicit none
+   !DEC$ ATTRIBUTES NO_ARG_CHECK :: base
+   !GCC$ ATTRIBUTES NO_ARG_CHECK :: base
+   !$PRAGMA IGNORE_TKR base
+   !DIR$ IGNORE_TKR base
+   !IBM* IGNORE_TKR base
+   OMPI_FORTRAN_IGNORE_TKR_TYPE :: base
+   INTEGER(MPI_ADDRESS_KIND), INTENT(IN) :: size
+   TYPE(MPI_Win), INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces.F90]: `win` should be an IN

File: ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces.F90 [hunk 5]
```diff
@@ -3219,6 +3221,48 @@ subroutine MPI_Win_create_f08(base,size,disp_unit,info,comm,win,ierror)
...
+subroutine MPI_Win_detach_f08(win,base,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
+   implicit none
+   !DEC$ ATTRIBUTES NO_ARG_CHECK :: base
+   !GCC$ ATTRIBUTES NO_ARG_CHECK :: base
+   !$PRAGMA IGNORE_TKR base
+   !DIR$ IGNORE_TKR base
+   !IBM* IGNORE_TKR base
+   OMPI_FORTRAN_IGNORE_TKR_TYPE :: base
+   TYPE(MPI_Win), INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces.F90]: `win` should be an IN

File: ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces.F90 [hunk 6]
```diff
@@ -3219,6 +3221,48 @@ subroutine MPI_Win_create_f08(base,size,disp_unit,info,comm,win,ierror)
...
+   OMPI_FORTRAN_IGNORE_TKR_TYPE :: base
+   INTEGER(MPI_ADDRESS_KIND), INTENT(IN) :: size
+   TYPE(MPI_Win), INTENT(OUT) :: win
+   INTEGER, OPTIONAL, INTENT(OUT) :: ierror
+end subroutine MPI_Win_attach_f08
+end interface  MPI_Win_attach
+
+interface  MPI_Win_detach
+subroutine MPI_Win_detach_f08(win,base,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces.F90:3253]: Don't need MPI_ADDRESS_KIND here

File: ompi/mpi/fortran/use-mpi-f08/profile/pwin_attach_f08.F90 [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+#include "ompi/mpi/fortran/configure-fortran-output.h"
+
+subroutine PMPI_Win_attach_f08(win,base,size,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
+   use :: mpi_f08, only : ompi_win_attach_f
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   INTEGER(MPI_ADDRESS_KIND), INTENT(IN) :: size
+   TYPE(MPI_Win), INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/profile/pwin_attach_f08.F90]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-f08/profile/pwin_detach_f08.F90 [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+! $COPYRIGHT$
+
+#include "ompi/mpi/fortran/configure-fortran-output.h"
+
+subroutine PMPI_Win_detach_f08(win,base,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
+   use :: mpi_f08, only : ompi_win_detach_f
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   TYPE(MPI_Win), INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/profile/pwin_detach_f08.F90]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-f08/profile/pwin_detach_f08.F90 [hunk 9]
```diff
@@ -0,0 +1,21 @@
+! -*- f90 -*-
+!
+! Copyright (c) 2015      Research Organization for Information Science
+!                         and Technology (RIST). All rights reserved.
+! $COPYRIGHT$
+
+#include "ompi/mpi/fortran/configure-fortran-output.h"
+
+subroutine PMPI_Win_detach_f08(win,base,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/profile/pwin_detach_f08.F90:10]: No need for MPI_ADDRESS_KIND here

File: ompi/mpi/fortran/use-mpi-f08/win_attach_f08.F90 [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+#include "ompi/mpi/fortran/configure-fortran-output.h"
+
+subroutine MPI_Win_attach_f08(win,base,size,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
+   use :: mpi_f08, only : ompi_win_attach_f
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   INTEGER(MPI_ADDRESS_KIND), INTENT(IN) :: size
+   TYPE(MPI_Win), INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/win_attach_f08.F90]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-f08/win_detach_f08.F90 [hunk 11]
```diff
@@ -0,0 +1,21 @@
+! -*- f90 -*-
+!
+! Copyright (c) 2015      Research Organization for Information Science
+!                         and Technology (RIST). All rights reserved.
+! $COPYRIGHT$
+
+#include "ompi/mpi/fortran/configure-fortran-output.h"
+
+subroutine MPI_Win_detach_f08(win,base,ierror)
+   use :: mpi_f08_types, only : MPI_Win, MPI_ADDRESS_KIND
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/win_detach_f08.F90:10]: No need for MPI_ADDRESS_KIND here

File: ompi/mpi/fortran/use-mpi-f08/win_detach_f08.F90 [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+   use :: mpi_f08, only : ompi_win_detach_f
+   implicit none
+   OMPI_FORTRAN_IGNORE_TKR_TYPE, INTENT(IN) :: base
+   TYPE(MPI_Win), INTENT(OUT) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/win_detach_f08.F90]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in [hunk 13]
```diff
@@ -7432,6 +7434,85 @@ end subroutine PMPI_Win_create
...
+
+
+interface MPI_Win_attach
+
+subroutine MPI_Win_attach(win, base, size, ierror)
+  include 'mpif-config.h'
+  @OMPI_FORTRAN_IGNORE_TKR_PREDECL@ base
+  @OMPI_FORTRAN_IGNORE_TKR_TYPE@, intent(in) :: base
+  integer(kind=MPI_ADDRESS_KIND), intent(in) :: size
+  integer, intent(out) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in [hunk 14]
```diff
@@ -7432,6 +7434,85 @@ end subroutine PMPI_Win_create
...
+end interface
+
+interface PMPI_Win_attach
+
+subroutine PMPI_Win_attach(win, base, size, ierror)
+  include 'mpif-config.h'
+  @OMPI_FORTRAN_IGNORE_TKR_PREDECL@ base
+  @OMPI_FORTRAN_IGNORE_TKR_TYPE@, intent(in) :: base
+  integer(kind=MPI_ADDRESS_KIND), intent(in) :: size
+  integer, intent(out) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in [hunk 15]
```diff
@@ -7432,6 +7434,85 @@ end subroutine PMPI_Win_create
...
+end interface
+
+
+interface MPI_Win_detach
+
+subroutine MPI_Win_detach(win, base, ierror)
+  include 'mpif-config.h'
+  @OMPI_FORTRAN_IGNORE_TKR_PREDECL@ base
+  @OMPI_FORTRAN_IGNORE_TKR_TYPE@, intent(in) :: base
+  integer, intent(out) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in]: `win` should be IN

File: ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in [hunk 16]
```diff
@@ -7432,6 +7434,85 @@ end subroutine PMPI_Win_create
...
+
+end interface
+
+interface PMPI_Win_detach
+
+subroutine PMPI_Win_detach(win, base, ierror)
+  include 'mpif-config.h'
+  @OMPI_FORTRAN_IGNORE_TKR_PREDECL@ base
+  @OMPI_FORTRAN_IGNORE_TKR_TYPE@, intent(in) :: base
+  integer, intent(out) :: win
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-ignore-tkr/mpi-ignore-tkr-interfaces.h.in]: `win` should be IN

**Comment by jsquyres** [comment]
Looks like the same small number of errors just got copy-n-pasted all the way through.  Should be easy to fix.

Thanks @ggouaillardet!

**Comment by ggouaillardet** [comment]
@jsquyres one commit was not cherry-picked from master.
i fixed more stuff in the master and cherry-picked into this PR.
if i remove

```
   use :: mpi_f08_types, only : MPI_ADDRESS_KIND
```

compilation fails

**Comment by ggouaillardet** [comment]
:bot:nolabel:pushed-back

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/925/ for details.

**Comment by ggouaillardet** [comment]
:bot:label:blocker

@jsquyres some mtt tests do not even build without this PR, so i just marked this as a blocker

**Comment by jsquyres** [comment]
Gilles - thanks for digging in to it. :+1:

**Pull request merged**